curator-client/src/main/java/org/apache/curator/HandleHolder.java

```diff
@@ -70,7 +61,16 @@ String  getConnectionString()
     String getNewConnectionString()
     {
         String helperConnectionString = (helper != null) ? helper.getConnectionString() : null;
-        return ((helperConnectionString != null) && !ensembleProvider.getConnectionString().equals(helperConnectionString)) ? helperConnectionString : null;
+        String ensembleProviderConnectionString = ensembleProvider.getConnectionString();
```

Comment:
Note: this is the major change

Reply:
I have a question. `helper` is `null` until `closeAndReset()` is called, and new connection string can be sensed only if `helper` is not `null`. Isn't it possible to have new connection string (from the `EnsembleTracker`) before `closeAndReset()` was ever called? If it is possible, then it seems like the new connection string won't be handled, since `helper` would be `null`.

Reply:
Thanks @shayshim I'll look into this possiblity

Reply:
The whole flow of how things interact is very confusing, but I don't think that this is an issue. `closeAndReset()` gets called from `ConnectionState.reset()` which gets called from the `ConnectionState.start()` method.  I'm not 100% sure there's not a race condition here, but from what I can see, it seems improbable.

Reply:
> whole flow of how things interact is very confusing

Too true. What do we do? Should we attempt a more aggressive change?

Reply:
I see, thanks @cammckenzie

Reply:
Maybe we should just merge this change and look at cleaning up this stuff in a future PR? I think it would be a pretty significant change.

Reply:
I'm OK with that. @shayshim ? Maybe you can both approve this PR when you're OK

Reply:
> I have a question. `helper` is `null` until `closeAndReset()` is called, and new connection string can be sensed only if `helper` is not `null`. Isn't it possible to have new connection string (from the `EnsembleTracker`) before `closeAndReset()` was ever called? If it is possible, then it seems like the new connection string won't be handled, since `helper` would be `null`.

`closeAndReset()` will cause `ensembleProvider.getConnectionString()` to get called anyway. Does that answer your question? I think the intent of the above code is to force-close the connection when there's a new connection string.

Reply:
> I'm OK with that. @shayshim ? Maybe you can both approve this PR when you're OK

I agree  - better to continue refactor in different PR, dedicated for that

Reply:
> `closeAndReset()` will cause `ensembleProvider.getConnectionString()` to get called anyway. Does that answer your question? I think the intent of the above code is to force-close the connection when there's a new connection string.
In short I wanted to make sure  that `closeAndReset()` (create `helper`) is called before any new connection string is set . 
As @cammckenzie said, it is called from `ConnectionState.start()`, which is called (eventually) from `CuratorFrameworkImpl.start()`, from line 338 using `client.start()` - before `ensembleTracker.start()` in line 353, which allows to set new connection strings. 
So any call to `getNewConnectionString()` will meet non `null` `helper`. 
In addition to that, `helper.getZooKeeper()` is also called eventually from 338, so `data.connectionString` is set as needed, before `getNewConnectionString()` can be called. 
So I think we have no issue here.